prog/mutation.go

```diff
@@ -269,6 +269,117 @@ func Minimize(p0 *Prog, callIndex0 int, pred func(*Prog, int) bool) (*Prog, int)
 		p0 = p
 		callIndex0 = callIndex
 	}
+
+	triedPaths := make(map[string]bool)
+
+	var rec func(p *Prog, call *Call, arg *Arg, path string) bool
+	rec = func(p *Prog, call *Call, arg *Arg, path string) bool {
+		switch typ := arg.Type.(type) {
+		case *sys.StructType:
+			for _, innerArg := range arg.Inner {
+				innerPath := fmt.Sprintf("%v-%v", path, innerArg.Type.Name())
+				if rec(p, call, innerArg, innerPath) {
+					return true
+				}
+			}
+		case *sys.UnionType:
+			if rec(p, call, arg.Option, fmt.Sprintf("%v-%v", path, arg.Option.Type.Name())) {
+				return true
+			}
+		case *sys.PtrType:
+			// TODO: remove optional ptrs
+			if arg.Res != nil {
+				return rec(p, call, arg.Res, fmt.Sprintf("%v-inner", path))
+			}
+			return false
+		case *sys.ArrayType:
+			for i, innerArg := range arg.Inner {
+				innerPath := fmt.Sprintf("%v-%v", path, i)
+				if (typ.Kind == sys.ArrayRangeLen && len(arg.Inner) > int(typ.RangeBegin)) ||
+				   (typ.Kind == sys.ArrayRandLen) {
+					if !triedPaths[innerPath] {
+						copy(arg.Inner[i:], arg.Inner[i+1:])
+						arg.Inner = arg.Inner[:len(arg.Inner)-1]
+						p.removeArg(call, innerArg)
+						assignSizesCall(call)
+
+						if pred(p, callIndex0) {
+							p0 = p
+						} else {
+							triedPaths[innerPath] = true
+						}
+
+						return true
+					}
+				}
+				if rec(p, call, innerArg, innerPath) {
+					return true
+				}
+			}
+		case *sys.IntType, *sys.FlagsType:
+			// TODO: try to set separate flags
+			if triedPaths[path] {
+				return false
+			}
+			if arg.Val == 0 {
+				triedPaths[path] = true
+				return true
+			}
+			arg.Val = 0
+			if pred(p, callIndex0) {
+				p0 = p
+			}
+			triedPaths[path] = true
+			return true
+		case *sys.BufferType:
+			if triedPaths[path] {
+				return false
+			}
+
+			var newLen uintptr
+
+			switch typ.Kind {
+			case sys.BufferBlobRand:
+				if len(arg.Data) == 0 {
+					triedPaths[path] = true
+					return false
+				}
+				newLen = uintptr(len(arg.Data)) / 2
+			case sys.BufferBlobRange:
+				if typ.RangeBegin == typ.RangeEnd || typ.RangeBegin == uintptr(len(arg.Data)) {
+					triedPaths[path] = true
+					return false
+				}
+				newLen = typ.RangeBegin + (uintptr(len(arg.Data)) - typ.RangeBegin) / 2
+			default:
+				triedPaths[path] = true
+				return false
+			}
+
+			arg.Data = arg.Data[:newLen]
```

Comment:
We need 12 iterations and 12 program clones to reduce a 4K buffer to empty buffer. We also never try to reduce buffer by less than half.
We need something like this:
```go
		case *sys.BufferType:
			if triedPaths[path] {
				return false
			}
			triedPaths[path] = true
			if typ.Kind != sys.BufferBlobRand && typ.Kind != sys.BufferBlobRange {
				return false
			}
			minLen := int(typ.RangeBegin)
			for step := len(arg.Data) - minLen; len(arg.Data) > minLen && step > 0; {
				if len(arg.Data) - step >= minLen {
					arg.Data = arg.Data[:len(arg.Data) - step]
					assignSizesCall(call)
					if pred(p, callIndex0) {
						continue
					}
					arg.Data = arg.Data[:len(arg.Data) + step]
					assignSizesCall(call)
				}
				step /= 2
			}
			p0 = p
```

We can also use similar approach to reset individual bits in ints.